gobblin-data-management/src/main/java/org/apache/gobblin/data/management/copy/iceberg/IcebergTableMetadataValidator.java

```diff
@@ -0,0 +1,127 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.gobblin.data.management.copy.iceberg;
+
+import java.util.Arrays;
+
+import org.apache.iceberg.PartitionSpec;
+import org.apache.iceberg.Schema;
+import org.apache.iceberg.TableMetadata;
+
+import com.google.common.annotations.VisibleForTesting;
+
+import lombok.extern.slf4j.Slf4j;
+
+/**
+ * Validator for Iceberg table metadata, ensuring that the source and destination tables have
+ * compatible schemas and partition specifications.
+ */
+@Slf4j
+public class IcebergTableMetadataValidator {
+
+  private IcebergTableMetadataValidator() {
+    // Do not instantiate
+  }
+
+  /**
+   * Validates the metadata of the source and destination Iceberg tables.
+   *
+   * @param srcTableMetadata  the metadata of the source table
+   * @param destTableMetadata the metadata of the destination table
+   * @throws IllegalArgumentException if the schemas or partition specifications do not match
```

Comment:
the `IllegalArgEx` javadoc describes:
```
Thrown to indicate that a method has been passed an illegal or inappropriate argument.
```
doesn't seem applicable here, as it's not so much about "args", but rather an incompatible structure.

at times we've used `IOException` to tunnel out specifics of the error

Reply:
My thought process was as it will be used from IcebergPartitionDatasetFinder so throwing IllegalArgumentException make somewhat sense in that context but yes can be changed to IOException

Reply:
Changed it to throw IOException